File: packages/core/src/research_engine/domain/citations.py
"""Citation occurrences and items — where a block leans on the corpus.

An occurrence is the `{{cite:<key>}}` marker's row: which block, which intent,
which placement. An item is one grounding of that marker: which span, which
typed quote at which tier, which identity. Two items over one span keep their
own wording and tier — decision 1, the same split the ledger's anchors use.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID  # noqa: TC003 - pydantic needs it at runtime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from research_engine.domain.works import Placement
from research_engine.domain.works_files import Intent


class CitationOccurrence(BaseModel):
    id: UUID
    citation_key: UUID
    block_id: UUID
    placement: Placement = Placement.INLINE
    intent: Intent = Intent.SOURCE
    note: str | None = None
    created_at: datetime


class OccurrenceDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    block_id: UUID
    citation_key: UUID
    placement: Placement = Placement.INLINE
    intent: Intent = Intent.SOURCE
    note: str | None = None


class CitationItem(BaseModel):
    occurrence_id: UUID
    position: int
    edition_id: UUID | None = None
    zotero_key: str | None = None
    source_span_id: UUID | None = None
    quoted_text: str | None = None
    verify_status: str | None = None
    verified_at: datetime | None = None
    locator: dict[str, Any] = Field(default_factory=dict)
    prefix: str | None = None
    suffix: str | None = None
    suppress_author: bool = False


class CitationItemDraft(BaseModel):
    """An item validates before the database does.

    Identity first: an item with neither `edition_id` nor `zotero_key` names
    no edition and is refused here, not by the check constraint. Then the
    quote: typed wording without a span is a claim without an address.
    """

    model_config = ConfigDict(extra="forbid")

    occurrence_id: UUID
    position: int = 0
    edition_id: UUID | None = None
    zotero_key: str | None = None
    source_span_id: UUID | None = None
    quoted_text: str | None = None
    verify_status: str | None = None
    locator: dict[str, Any] = Field(default_factory=dict)
    prefix: str | None = None
    suffix: str | None = None
    suppress_author: bool = False

    @model_validator(mode="after")
    def _names_an_edition(self) -> CitationItemDraft:
        if self.edition_id is None and self.zotero_key is None:
            raise ValueError("an item names its edition: edition_id or zotero_key")
        return self

    @model_validator(mode="after")
    def _quote_needs_a_span(self) -> CitationItemDraft:
        if self.quoted_text is not None and self.source_span_id is None:
            raise ValueError("quoted_text without source_span_id is unaddressed")
        return self


class BlockCitations(BaseModel):
    """One occurrence with everything grounding it, for export and validate."""

    occurrence: CitationOccurrence
    items: list[CitationItem] = Field(default_factory=list)

File: crates/marginalia-types/src/citations.rs
//! Citation occurrences and items mirroring `domain/citations.py`.
//!
//! An occurrence is the `{{cite:<key>}}` marker's row: which block, which
//! intent, which placement. An item is one grounding of that marker.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::errors::{Error, Result};
use crate::works::Placement;
use crate::works_files::Intent;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CitationOccurrence {
    pub id: Uuid,
    pub citation_key: Uuid,
    pub block_id: Uuid,
    #[serde(default)]
    pub placement: Placement,
    #[serde(default)]
    pub intent: Intent,
    #[serde(default)]
    pub note: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OccurrenceDraft {
    pub block_id: Uuid,
    pub citation_key: Uuid,
    #[serde(default)]
    pub placement: Placement,
    #[serde(default)]
    pub intent: Intent,
    #[serde(default)]
    pub note: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CitationItem {
    pub occurrence_id: Uuid,
    pub position: i64,
    #[serde(default)]
    pub edition_id: Option<Uuid>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub source_span_id: Option<Uuid>,
    #[serde(default)]
    pub quoted_text: Option<String>,
    #[serde(default)]
    pub verify_status: Option<String>,
    #[serde(default)]
    pub verified_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub locator: Map<String, Value>,
    #[serde(default)]
    pub prefix: Option<String>,
    #[serde(default)]
    pub suffix: Option<String>,
    #[serde(default)]
    pub suppress_author: bool,
}

/// An item validates before the database does: identity first, then the quote.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CitationItemDraft {
    pub occurrence_id: Uuid,
    #[serde(default)]
    pub position: i64,
    #[serde(default)]
    pub edition_id: Option<Uuid>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub source_span_id: Option<Uuid>,
    #[serde(default)]
    pub quoted_text: Option<String>,
    #[serde(default)]
    pub verify_status: Option<String>,
    #[serde(default)]
    pub locator: Map<String, Value>,
    #[serde(default)]
    pub prefix: Option<String>,
    #[serde(default)]
    pub suffix: Option<String>,
    #[serde(default)]
    pub suppress_author: bool,
}

impl CitationItemDraft {
    /// The `_names_an_edition` and `_quote_needs_a_span` model validators.
    pub fn validate(&self) -> Result<()> {
        if self.edition_id.is_none() && self.edition_key.is_none() {
            return Err(Error::Validation(
                "an item names its edition: edition_id or edition_key".to_owned(),
            ));
        }
        if self.quoted_text.is_some() && self.source_span_id.is_none() {
            return Err(Error::Validation(
                "quoted_text without source_span_id is unaddressed".to_owned(),
            ));
        }
        Ok(())
    }
}

/// One occurrence with everything grounding it, for export and validate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BlockCitations {
    pub occurrence: CitationOccurrence,
    #[serde(default)]
    pub items: Vec<CitationItem>,
}
